### PR TITLE
Add device_doctor ios management logic

### PR DESCRIPTION
**Pull request by keyonghan**
This PR implements basic health check:  device accessible, and recovery: restart and dismiss dialogs.

Related issue: https://github.com/flutter/flutter/issues/66193

### CODE REVIEW

File: device_doctor/bin/main.dart [hunk 1]
```diff
@@ -5,6 +5,7 @@
 import 'package:args/args.dart';
 
 import 'package:device_doctor/src/device.dart';
+import 'package:device_doctor/src/ios_device.dart';
```
Comment (CaseyHillers) [inline device_doctor/bin/main.dart:8]: Consider exporting what's needed from src, similar to how app_dart/lib/cocoon_service.dart works.

https://dart.dev/guides/language/effective-dart/usage#do-use-relative-paths-when-importing-libraries-within-your-own-packages-lib-directory

File: device_doctor/bin/main.dart [hunk 2]
```diff
@@ -14,9 +15,15 @@ const List<String> supportedDeviceOS = <String>['ios', 'android'];
...
+String _action;
 String _deviceOS;
 
-/// Manage healthcheck and recovery for devices.
+/// Manage `healthcheck` and `recovery` for devices.
+///
+/// For `healthcheck`, if no device is found or any health check fails an stderr will be logged,
+/// and an exception will be thrown.
+///
+/// For `recovery`, device will do cleanup, reboot, and then dismiss any dialog for `ios` device.
```
Comment (CaseyHillers) [inline device_doctor/bin/main.dart:26]: > then dismiss any dialog for `ios` device

Could this be generalized to initialize device? If there's future dialogs that need to be closed for android, this doc would fall out of date.

File: device_doctor/lib/src/health.dart [hunk 3]
```diff
@@ -2,20 +2,72 @@
...
+      command.add("PROVISIONING_PROFILE_SPECIFIER=${pl.environment['FLUTTER_XCODE_PROVISIONING_PROFILE_SPECIFIER']}");
+    }
+    Process proc = await pm.start(command, workingDirectory: dialogDir.path);
+    logger.info('Executing: $command');
+    // Discards stdout and stderr as they are too large.
+    await proc.stdout.drain<Object>();
+    await proc.stderr.drain<Object>();
+    int exitCode = await proc.exitCode;
+    if (exitCode != 0) {
+      fail('Command "xcrun xcodebuild -project infra-dialog..." failed with exit code $exitCode.');
```
Comment (CaseyHillers) [inline device_doctor/lib/src/health.dart:49]: The command should be moved into a variable so changes made to one place propagate to all calls.

File: device_doctor/lib/src/health.dart [hunk 4]
```diff
@@ -26,3 +78,25 @@ class HealthCheckResult {
...
 }
+
+/// Check healthiness for discovered devices.
+///
+/// If any failed check, an explanation message will be sent to stdout and
+/// an exception will be thrown.
+Future<void> healthcheck(Map<String, List<HealthCheckResult>> deviceChecks) async {
+  if (deviceChecks.isEmpty) {
+    stderr.writeln('No healthy device is available');
+    throw StateError('No healthy device is available');
```
Comment (CaseyHillers) [inline device_doctor/lib/src/health.dart:89]: The bot config should already write the error when it is thrown, why are we also logging it?

File: device_doctor/test/src/health_test.dart [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    test('failed', () async {
+      Process proc = FakeProcess(123);
+      when(pm.start(any, workingDirectory: anyNamed('workingDirectory'))).thenAnswer((_) => Future.value(proc));
+
+      expect(
+        closeIosDialog(pm: pm),
+        throwsA(TypeMatcher<BuildFailedError>()),
+      );
+    });
+  });
```
Comment (CaseyHillers) [inline device_doctor/test/src/health_test.dart:55]: What about a test where the tool cannot be found?

File: device_doctor/lib/src/utils.dart [hunk 6]
```diff
@@ -18,3 +21,33 @@ class BuildFailedError extends Error {
   @override
   String toString() => message;
 }
+
+/// Creates a directory from the given path, or multiple path parts by joining
+/// them using OS-specific file path separator.
+Directory dir(String thePath,
+    [String part2, String part3, String part4, String part5, String part6, String part7, String part8]) {
```
Comment (CaseyHillers) [inline device_doctor/lib/src/utils.dart:28]: Spread operator would clean this up.

https://dart.dev/guides/language/language-tour#spread-operator

File: device_doctor/test/src/health_test.dart (hunk 5 above)
Comment (keyonghan) [reply device_doctor/test/src/health_test.dart:55]: Added.

File: device_doctor/lib/src/utils.dart (hunk 6 above)
Comment (keyonghan) [reply device_doctor/lib/src/utils.dart:28]: This is to align with: https://github.com/dart-lang/path/blob/master/lib/path.dart#L257

File: device_doctor/lib/src/health.dart (hunk 4 above)
Comment (keyonghan) [reply device_doctor/lib/src/health.dart:89]: Added for testing. Removed.

File: device_doctor/lib/src/health.dart (hunk 3 above)
Comment (keyonghan) [reply device_doctor/lib/src/health.dart]: `command` is already there. Have included variable `command` in the message.

File: device_doctor/bin/main.dart (hunk 2 above)
Comment (keyonghan) [reply device_doctor/bin/main.dart]: Updated to make it more general.

File: device_doctor/bin/main.dart (hunk 1 above)
Comment (keyonghan) [reply device_doctor/bin/main.dart]: Updated.

File: device_doctor/lib/src/health.dart [hunk 7]
```diff
@@ -17,10 +17,11 @@ import 'utils.dart';
 Future<HealthCheckResult> closeIosDialog(
     {ProcessManager pm = const LocalProcessManager(),
     String deviceId,
-    platform.Platform pl = const platform.LocalPlatform()}) async {
-  Directory dialogDir = dir(path.dirname(Platform.script.path), 'tool', 'infra-dialog');
+    platform.Platform pl = const platform.LocalPlatform(),
+    String infraDialog = 'infra-dialog'}) async {
```
Comment (CaseyHillers) [inline device_doctor/lib/src/health.dart:21]: nit: if you add a comma after `infra-dialog',` it will format the args to each be on their own line

**Review by CaseyHillers — APPROVED** [review]
LGTM!

File: device_doctor/lib/src/health.dart (hunk 7 above)
Comment (keyonghan) [reply device_doctor/lib/src/health.dart:21]: reformatted. Thanks.

**Pull request merged**